src/repositories/gitlab.rs

```diff
@@ -0,0 +1,222 @@
+use crate::error::Result;
+use chrono::{DateTime, Utc};
+use log::warn;
+use reqwest::{
+    blocking::Client as HttpClient,
+    header::{HeaderMap, HeaderValue, ACCEPT, AUTHORIZATION, USER_AGENT},
+};
+use serde::Deserialize;
+use std::collections::HashSet;
+
+use crate::repositories::{
+    FetchRepositoriesResult, Repository, RepositoryForge, RepositoryName, APP_USER_AGENT,
+};
+
+const GRAPHQL_UPDATE: &str = "query($ids: [ID!]!) {
+    projects(ids: $ids) {
+        nodes {
+            id
+            nameWithNamespace
+            lastActivityAt
+            description
+            starCount
+            forksCount
+            openIssuesCount
+        }
+    }
+}";
+
+const GRAPHQL_SINGLE: &str = "query($fullPath: ID!) {
+    project(fullPath: $fullPath) {
+        id
+        nameWithNamespace
+        lastActivityAt
+        description
+        starCount
+        forksCount
+        openIssuesCount
+    }
+}";
+
+pub struct GitLab {
+    client: HttpClient,
+    host: &'static str,
+}
+
+impl GitLab {
+    pub fn new(host: &'static str, access_token: &Option<String>) -> Result<Option<Self>> {
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static(APP_USER_AGENT));
+        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
+
+        if let Some(token) = access_token {
+            headers.insert(
+                AUTHORIZATION,
+                HeaderValue::from_str(&format!("Bearer {}", token))?,
+            );
+        } else {
+            warn!(
+                "will try to retrieve `{}` stats without token since none was provided",
+                host
+            );
+        }
+
+        let client = HttpClient::builder().default_headers(headers).build()?;
+        Ok(Some(GitLab { client, host }))
+    }
+}
+
+impl RepositoryForge for GitLab {
+    fn host(&self) -> &str {
+        self.host
+    }
+
+    fn icon(&self) -> &'static str {
+        "gitlab"
+    }
+
+    fn chunk_size(&self) -> usize {
+        5
```

Comment:
Hmm, GitLab supports at most 5 fetched projects per query? That seems low.

Reply:
It'll require "bigger" tests. I wasn't able to reach the limit.